In [1]:
import pandas as pd
import numpy as np

In [2]:
#df = pd.read_excel('D:/Docs_copy/Capstone/Resource Data/Amazon EC2 Instance Comparison.xlsx')

In [3]:
df=pd.read_csv('D:/Docs_copy/Capstone/Resource Data/Amazon EC2 Instance Comparison1.csv')

In [5]:
df.head()

Name     API Name  Memory Compute Units (ECU)  Cores  \
0                 R6G Medium   r6g.medium    8.00             0 units      1   
1                C6GD Medium  c6gd.medium    2.00             0 units      1   
2  M3 General Purpose Medium    m3.medium    3.75             3 units      1   
3                 M6G Medium   m6g.medium    4.00             0 units      1   
4  M1 General Purpose Medium    m1.medium    3.75             2 units      1   

   GPUs GPU model  GPU memory  CUDA Compute Capability  FPGAs  ...  \
0     0      None           0                      0.0      0  ...   
1     0      None           0                      0.0      0  ...   
2     0      None           0                      0.0      0  ...   
3     0      None           0                      0.0      0  ...   
4     0      None           0                      0.0      0  ...   

  Windows SQL Ent On Demand cost Windows SQL Ent Reserved cost  \
0                            NaN                           NaN   
1                            NaN                           NaN   
2                            NaN                           NaN   
3                            NaN                           NaN   
4                            NaN                           NaN   

   Linux SQL Web On Demand cost Linux SQL Web Reserved cost  \
0                           NaN                         NaN   
1                           NaN                         NaN   
2                        $0.14                       $0.11    
3                           NaN                         NaN   
4                           NaN                         NaN   

  Linux SQL Std On Demand cost Linux SQL Std Reserved cost  \
0                          NaN                         NaN   
1                          NaN                         NaN   
2                       $0.55                       $0.52    
3                          NaN                         NaN   
4                          NaN                         NaN   

  Linux SQL Ent On Demand cost Linux SQL Ent Reserved cost  \
0                          NaN                         NaN   
1                          NaN                         NaN   
2                          NaN                         NaN   
3                          NaN                         NaN   
4                          NaN                         NaN   

  EBS Optimized surcharge EMR cost  
0                       0      NaN  
1                       0      NaN  
2                     NaN      NaN  
3                       0      NaN  
4                     NaN   $0.02   

[5 rows x 56 columns]

In [6]:
df.columns

Index(['Name', 'API Name', 'Memory', 'Compute Units (ECU)', 'Cores', 'GPUs',
       'GPU model', 'GPU memory', 'CUDA Compute Capability', 'FPGAs',
       'ECU per vCPU', 'Physical Processor', 'Clock Speed(GHz)', 'Intel AVX',
       'Intel AVX2', 'Intel AVX-512', 'Intel Turbo', 'Instance Storage',
       'Instance Storage: already warmed-up',
       'Instance Storage: SSD TRIM Support', 'Arch', 'Network Performance',
       'EBS Optimized: Max Bandwidth', 'EBS Optimized: Max Throughput (128K)',
       'EBS Optimized: Max IOPS (16K)', 'EBS Exposed as NVMe', 'Max IPs',
       'Max ENIs', 'Enhanced Networking', 'VPC Only', 'IPv6 Support',
       'Placement Group Support', 'Linux Virtualization', 'On EMR',
       'Linux On Demand cost', 'Linux Reserved cost', 'RHEL On Demand cost',
       'RHEL Reserved cost', 'SLES On Demand cost', 'SLES Reserved cost',
       'Windows On Demand cost', 'Windows Reserved cost',
       'Windows SQL Web On Demand cost', 'Windows SQL Web Reserved cost',
      

In [7]:
cost_df=df[['Name','API Name','Memory','Cores','Physical Processor','Clock Speed(GHz)','Linux On Demand cost', 'Linux Reserved cost']]

In [8]:
p1 = pd.read_csv('D:/Docs_copy/Capstone/Resource Data/Forecasted Data.csv')

In [9]:
p1.head()

Model  CPU_MAE  CPU_RMSE  MPU_MAE  MPU_RMSE  VM        Timestamp  \
0  AutoRegressive   0.0107   0.03212  5.56943   2.35996   3  9/29/2013 22:03   
1  AutoRegressive   0.0107   0.03212  5.56943   2.35996   3  9/29/2013 22:08   
2  AutoRegressive   0.0107   0.03212  5.56943   2.35996   3  9/29/2013 22:13   
3  AutoRegressive   0.0107   0.03212  5.56943   2.35996   3  9/29/2013 22:18   
4  AutoRegressive   0.0107   0.03212  5.56943   2.35996   3  9/29/2013 22:23   

   CPU cores  CPU capacity provisioned [MHZ]  predicted_cpu_usage [%]  \
0          2                     2599.999304                 0.994962   
1          2                     2599.999304                 0.994962   
2          2                     2599.999304                 0.994962   
3          2                     2599.999304                 0.994962   
4          2                     2599.999304                 0.994962   

   CPU usage [MHZ]  Memory capacity provisioned [KB]  \
0        25.869017                            702948   
1        25.869017                            702948   
2        25.869017                            702948   
3        25.869017                            702948   
4        25.869017                            702948   

   predicted_memory_usage [%]  Memory usage [KB]  
0                   10.515294        73917.05093  
1                   10.515294        73917.05093  
2                   10.515294        73917.05093  
3                   10.515294        73917.05093  
4                   10.515294        73917.05093

In [10]:
grouped_data = p1.groupby(['Model','VM']).agg({
                                                'CPU_MAE':'mean',
                                                'CPU_RMSE':'mean',
                                                'MPU_MAE' : 'mean',
                                                'MPU_RMSE' : 'mean',
                                                'CPU cores':'max',
                                                'CPU capacity provisioned [MHZ]':'mean',
                                                'predicted_cpu_usage [%]':'max',
                                                'CPU usage [MHZ]' : 'max',
                                                'Memory capacity provisioned [KB]' : 'mean',
                                                'predicted_memory_usage [%]':'max',
                                                'Memory usage [KB]':'max'
                                            
})

In [11]:
forecasted_grouped_data=grouped_data.reset_index()

In [12]:
forecasted_grouped_data['CPU Provisoned']=[2.5,2.5,3,2.5]
forecasted_grouped_data['Memory Provisoned']=[8,8,8,8]

In [13]:
forecasted_grouped_data

Model   VM  CPU_MAE  CPU_RMSE  MPU_MAE  MPU_RMSE  CPU cores  \
0    AutoRegressive    3   0.0107   0.03212  5.56943   2.35996          2   
1  AutoRegressive_2  198   0.4340   0.59020  1.28500   1.64150          2   
2        FB Prophet    1   1.6741   2.40698  2.97718   3.98274          2   
3    Moving Average  198   0.4406   0.59730  1.28500   1.64150          2   

   CPU capacity provisioned [MHZ]  predicted_cpu_usage [%]  CPU usage [MHZ]  \
0                     2599.999304                 0.994962        25.869017   
1                     5199.999272                50.495910      2625.786931   
2                     5851.998900                 9.223119       539.736850   
3                     5199.999272                50.492804      2625.625428   

   Memory capacity provisioned [KB]  predicted_memory_usage [%]  \
0                            702948                   10.515294   
1                           8257536                   17.294269   
2                           8218624                   13.752128   
3                           8257536                   17.295391   

   Memory usage [KB]  CPU Provisoned  Memory Provisoned  
0       7.391705e+04             2.5                  8  
1       1.428081e+06             2.5                  8  
2       1.130236e+06             3.0                  8  
3       1.428173e+06             2.5                  8

In [14]:
cost_df.head()

Name     API Name  Memory  Cores  \
0                 R6G Medium   r6g.medium    8.00      1   
1                C6GD Medium  c6gd.medium    2.00      1   
2  M3 General Purpose Medium    m3.medium    3.75      1   
3                 M6G Medium   m6g.medium    4.00      1   
4  M1 General Purpose Medium    m1.medium    3.75      1   

                                Physical Processor  Clock Speed(GHz)  \
0                          AWS Graviton2 Processor               2.5   
1                          AWS Graviton2 Processor               2.5   
2  Intel Xeon E5-2670 v2 (Ivy Bridge/Sandy Bridge)               2.5   
3                          AWS Graviton2 Processor               NaN   
4                                Intel Xeon Family               NaN   

   Linux On Demand cost  Linux Reserved cost  
0                  0.05                 0.03  
1                  0.04                 0.02  
2                  0.07                 0.05  
3                  0.04                 0.02  
4                  0.09                 0.06

In [15]:
filtered_vm_cpu2=cost_df[(cost_df['Memory']==8.00) & (cost_df['Clock Speed(GHz)']==2.5) & (cost_df['Cores']==2)]
print(filtered_vm_cpu2)

           Name    API Name  Memory  Cores       Physical Processor  \
17   M5AD Large  m5ad.large     8.0      2            AMD EPYC 7571   
21   M6GD Large  m6gd.large     8.0      2  AWS Graviton2 Processor   
32    M5A Large   m5a.large     8.0      2            AMD EPYC 7571   
340   T4G Large   t4g.large     8.0      2  AWS Graviton2 Processor   
341   T3A Large   t3a.large     8.0      2            AMD EPYC 7571   

     Clock Speed(GHz)  Linux On Demand cost  Linux Reserved cost  
17                2.5                  0.10                 0.07  
21                2.5                  0.09                 0.06  
32                2.5                  0.09                 0.05  
340               2.5                  0.07                 0.04  
341               2.5                  0.08                 0.05  


In [16]:
filtered_vm_cpu3=cost_df[(cost_df['Memory']==8.00) & (cost_df['Clock Speed(GHz)']==3.1)& (cost_df['Cores']==2)]
filtered_vm_cpu3

Name    API Name  Memory  Cores  \
18                 M5DN Large  m5dn.large     8.0      2   
22                  M5N Large   m5n.large     8.0      2   
40   M5 General Purpose Large    m5.large     8.0      2   
41   M5 General Purpose Large   m5d.large     8.0      2   
342                  T3 Large    t3.large     8.0      2   

                          Physical Processor  Clock Speed(GHz)  \
18   Intel Xeon Platinum 8259 (Cascade Lake)               3.1   
22   Intel Xeon Platinum 8259 (Cascade Lake)               3.1   
40        Intel Xeon Platinum 8175 (Skylake)               3.1   
41        Intel Xeon Platinum 8175 (Skylake)               3.1   
342                 Intel Skylake E5 2686 v5               3.1   

     Linux On Demand cost  Linux Reserved cost  
18                   0.14                 0.09  
22                   0.12                 0.08  
40                   0.10                 0.06  
41                   0.11                 0.07  
342                  0.08                 0.05

## Calculate based on Multiplying cost with 2160(no of hours) for 3 months

In [17]:
forecasted_grouped_data['Cost of OnDemand Linux Service']=[222.48,194.4,293.76,194.4]

In [18]:
forecasted_grouped_data['Cost of Reserved Linux Service']=[140.4,123.12,172.8,123.12]

In [19]:
forecasted_grouped_data['Forecasted Cost of OnDemand Linux Service for 1 Year']=[889.92,777.6,1175.04,777.6]

In [20]:
forecasted_grouped_data['Forecasted Cost of Reserved Linux Service for 1 Year']=[561.6,492.48,691.2,492.48]

In [21]:
forecasted_grouped_data

Model   VM  CPU_MAE  CPU_RMSE  MPU_MAE  MPU_RMSE  CPU cores  \
0    AutoRegressive    3   0.0107   0.03212  5.56943   2.35996          2   
1  AutoRegressive_2  198   0.4340   0.59020  1.28500   1.64150          2   
2        FB Prophet    1   1.6741   2.40698  2.97718   3.98274          2   
3    Moving Average  198   0.4406   0.59730  1.28500   1.64150          2   

   CPU capacity provisioned [MHZ]  predicted_cpu_usage [%]  CPU usage [MHZ]  \
0                     2599.999304                 0.994962        25.869017   
1                     5199.999272                50.495910      2625.786931   
2                     5851.998900                 9.223119       539.736850   
3                     5199.999272                50.492804      2625.625428   

   Memory capacity provisioned [KB]  predicted_memory_usage [%]  \
0                            702948                   10.515294   
1                           8257536                   17.294269   
2                           8218624                   13.752128   
3                           8257536                   17.295391   

   Memory usage [KB]  CPU Provisoned  Memory Provisoned  \
0       7.391705e+04             2.5                  8   
1       1.428081e+06             2.5                  8   
2       1.130236e+06             3.0                  8   
3       1.428173e+06             2.5                  8   

   Cost of OnDemand Linux Service  Cost of Reserved Linux Service  \
0                          222.48                          140.40   
1                          194.40                          123.12   
2                          293.76                          172.80   
3                          194.40                          123.12   

   Forecasted Cost of OnDemand Linux Service for 1 Year  \
0                                             889.92      
1                                             777.60      
2                                            1175.04      
3                                             777.60      

   Forecasted Cost of Reserved Linux Service for 1 Year  
0                                             561.60     
1                                             492.48     
2                                             691.20     
3                                             492.48

## Taking only Memory into Consideration, Filtering all memory less than 8 Gb


## For VM 3 Highest predicted memory usage % is 10.515 and cpu % is 0.9, Hence we are using less that

In [22]:
VM_3=cost_df[(cost_df['Memory']<8) & (cost_df['Clock Speed(GHz)']<2.5) & (cost_df['Cores']<2)]
VM_3.loc[:,'VM']=3
VM_3.loc[:,'Cost for Linux On Demand for 1 Year']=VM_3['Linux On Demand cost']*8640
VM_3.loc[:,'Cost for Linux Reserved for 1 Year']=VM_3['Linux Reserved cost']*8640

C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1596: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = _infer_fill_value(value)
C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)
C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1743: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value 

## For VM 198 Highest predicted memory usage % is 17 and cpu % is 50, Hence we are using less than or equal

In [23]:
VM_198=cost_df[(cost_df['Memory']<=8) & (cost_df['Clock Speed(GHz)']<=2.5) & (cost_df['Cores']==2)]
VM_198.loc[:,'VM']=198
VM_198.loc[:,'Cost for Linux On Demand for 1 Year']=VM_198['Linux On Demand cost']*8640
VM_198.loc[:,'Cost for Linux Reserved for 1 Year']=VM_198['Linux Reserved cost']*8640

C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1596: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = _infer_fill_value(value)
C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)
C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1743: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value 

## For VM 1 Highest predicted memory usage % is 13 and cpu % is 9, Hence we are using less than or equal for 1 year

In [24]:
VM_1=cost_df[(cost_df['Memory']<=8) & (cost_df['Clock Speed(GHz)']<2.5) & (cost_df['Cores']<=2)]
VM_1.loc[:,'VM']=1
VM_1.loc[:,'Cost for Linux On Demand for 1 Year']=VM_1['Linux On Demand cost']*8640
VM_1.loc[:,'Cost for Linux Reserved for 1 Year']=VM_1['Linux Reserved cost']*8640

C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1596: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = _infer_fill_value(value)
C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)
C:\Users\chat2\Anaconda3\lib\site-packages\pandas\core\indexing.py:1743: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value 

In [25]:
VM_1

Name   API Name  Memory  Cores  \
8                  A1 Medium  a1.medium     2.0      1   
13  M4 General Purpose Large   m4.large     8.0      2   
34                  A1 Large   a1.large     4.0      2   

                 Physical Processor  Clock Speed(GHz)  Linux On Demand cost  \
8            AWS Graviton Processor               2.3                  0.03   
13  Intel Xeon E5-2676 v3 (Haswell)               2.4                  0.10   
34           AWS Graviton Processor               2.3                  0.05   

    Linux Reserved cost  VM  Cost for Linux On Demand for 1 Year  \
8                  0.02   1                                259.2   
13                 0.06   1                                864.0   
34                 0.03   1                                432.0   

    Cost for Linux Reserved for 1 Year  
8                                172.8  
13                               518.4  
34                               259.2

In [26]:
frames = [VM_1, VM_3, VM_198]

recommendations = pd.concat(frames)

In [27]:
recommendations

Name    API Name  Memory  Cores  \
8                   A1 Medium   a1.medium     2.0      1   
13   M4 General Purpose Large    m4.large     8.0      2   
34                   A1 Large    a1.large     4.0      2   
8                   A1 Medium   a1.medium     2.0      1   
13   M4 General Purpose Large    m4.large     8.0      2   
17                 M5AD Large  m5ad.large     8.0      2   
19   M3 General Purpose Large    m3.large     7.5      2   
21                 M6GD Large  m6gd.large     8.0      2   
24                 C6GD Large  c6gd.large     4.0      2   
32                  M5A Large   m5a.large     8.0      2   
34                   A1 Large    a1.large     4.0      2   
35                  C6G Large   c6g.large     4.0      2   
326                  T3A Nano    t3a.nano     0.5      2   
327                  T4G Nano    t4g.nano     0.5      2   
329                 T4G Micro   t4g.micro     1.0      2   
330                 T3A Micro   t3a.micro     1.0      2   
332                T3A Medium  t3a.medium     4.0      2   
333                 T4G Small   t4g.small     2.0      2   
334                 T3A Small   t3a.small     2.0      2   
336                T4G Medium  t4g.medium     4.0      2   
340                 T4G Large   t4g.large     8.0      2   
341                 T3A Large   t3a.large     8.0      2   

                                  Physical Processor  Clock Speed(GHz)  \
8                             AWS Graviton Processor               2.3   
13                   Intel Xeon E5-2676 v3 (Haswell)               2.4   
34                            AWS Graviton Processor               2.3   
8                             AWS Graviton Processor               2.3   
13                   Intel Xeon E5-2676 v3 (Haswell)               2.4   
17                                     AMD EPYC 7571               2.5   
19   Intel Xeon E5-2670 v2 (Ivy Bridge/Sandy Bridge)               2.5   
21                           AWS Graviton2 Processor               2.5   
24                           AWS Graviton2 Processor               2.5   
32                                     AMD EPYC 7571               2.5   
34                            AWS Graviton Processor               2.3   
35                           AWS Graviton2 Processor               2.5   
326                                    AMD EPYC 7571               2.5   
327                          AWS Graviton2 Processor               2.5   
329                          AWS Graviton2 Processor               2.5   
330                                    AMD EPYC 7571               2.5   
332                                    AMD EPYC 7571               2.5   
333                          AWS Graviton2 Processor               2.5   
334                                    AMD EPYC 7571               2.5   
336                          AWS Graviton2 Processor               2.5   
340                          AWS Graviton2 Processor               2.5   
341                                    AMD EPYC 7571               2.5   

     Linux On Demand cost  Linux Reserved cost   VM  \
8                    0.03                 0.02    1   
13                   0.10                 0.06    1   
34                   0.05                 0.03    1   
8                    0.03                 0.02    3   
13                   0.10                 0.06  198   
17                   0.10                 0.07  198   
19                   0.13                 0.10  198   
21                   0.09                 0.06  198   
24                   0.08                 0.05  198   
32                   0.09                 0.05  198   
34                   0.05                 0.03  198   
35                   0.07                 0.04  198   
326                  0.00                 0.00  198   
327                  0.00                 0.00  198   
329                  0.01                 0.01  198   
330                  0.01                 0.01  198   
332   

In [28]:
#recommendations.to_csv('D:/Docs_copy/Capstone/Resource Data/recommendations.csv')

In [29]:
recommendations.head()

Name   API Name  Memory  Cores  \
8                  A1 Medium  a1.medium     2.0      1   
13  M4 General Purpose Large   m4.large     8.0      2   
34                  A1 Large   a1.large     4.0      2   
8                  A1 Medium  a1.medium     2.0      1   
13  M4 General Purpose Large   m4.large     8.0      2   

                 Physical Processor  Clock Speed(GHz)  Linux On Demand cost  \
8            AWS Graviton Processor               2.3                  0.03   
13  Intel Xeon E5-2676 v3 (Haswell)               2.4                  0.10   
34           AWS Graviton Processor               2.3                  0.05   
8            AWS Graviton Processor               2.3                  0.03   
13  Intel Xeon E5-2676 v3 (Haswell)               2.4                  0.10   

    Linux Reserved cost   VM  Cost for Linux On Demand for 1 Year  \
8                  0.02    1                                259.2   
13                 0.06    1                                864.0   
34                 0.03    1                                432.0   
8                  0.02    3                                259.2   
13                 0.06  198                                864.0   

    Cost for Linux Reserved for 1 Year  
8                                172.8  
13                               518.4  
34                               259.2  
8                                172.8  
13                               518.4

In [30]:
recommendations_1=recommendations[recommendations['VM']==1]
recommendations_3=recommendations[recommendations['VM']==3]
recommendations_198=recommendations[recommendations['VM']==198]
recommendations_198

Name    API Name  Memory  Cores  \
13   M4 General Purpose Large    m4.large     8.0      2   
17                 M5AD Large  m5ad.large     8.0      2   
19   M3 General Purpose Large    m3.large     7.5      2   
21                 M6GD Large  m6gd.large     8.0      2   
24                 C6GD Large  c6gd.large     4.0      2   
32                  M5A Large   m5a.large     8.0      2   
34                   A1 Large    a1.large     4.0      2   
35                  C6G Large   c6g.large     4.0      2   
326                  T3A Nano    t3a.nano     0.5      2   
327                  T4G Nano    t4g.nano     0.5      2   
329                 T4G Micro   t4g.micro     1.0      2   
330                 T3A Micro   t3a.micro     1.0      2   
332                T3A Medium  t3a.medium     4.0      2   
333                 T4G Small   t4g.small     2.0      2   
334                 T3A Small   t3a.small     2.0      2   
336                T4G Medium  t4g.medium     4.0      2   
340                 T4G Large   t4g.large     8.0      2   
341                 T3A Large   t3a.large     8.0      2   

                                  Physical Processor  Clock Speed(GHz)  \
13                   Intel Xeon E5-2676 v3 (Haswell)               2.4   
17                                     AMD EPYC 7571               2.5   
19   Intel Xeon E5-2670 v2 (Ivy Bridge/Sandy Bridge)               2.5   
21                           AWS Graviton2 Processor               2.5   
24                           AWS Graviton2 Processor               2.5   
32                                     AMD EPYC 7571               2.5   
34                            AWS Graviton Processor               2.3   
35                           AWS Graviton2 Processor               2.5   
326                                    AMD EPYC 7571               2.5   
327                          AWS Graviton2 Processor               2.5   
329                          AWS Graviton2 Processor               2.5   
330                                    AMD EPYC 7571               2.5   
332                                    AMD EPYC 7571               2.5   
333                          AWS Graviton2 Processor               2.5   
334                                    AMD EPYC 7571               2.5   
336                          AWS Graviton2 Processor               2.5   
340                          AWS Graviton2 Processor               2.5   
341                                    AMD EPYC 7571               2.5   

     Linux On Demand cost  Linux Reserved cost   VM  \
13                   0.10                 0.06  198   
17                   0.10                 0.07  198   
19                   0.13                 0.10  198   
21                   0.09                 0.06  198   
24                   0.08                 0.05  198   
32                   0.09                 0.05  198   
34                   0.05                 0.03  198   
35                   0.07                 0.04  198   
326                  0.00                 0.00  198   
327                  0.00                 0.00  198   
329                  0.01                 0.01  198   
330                  0.01                 0.01  198   
332                  0.04                 0.02  198   
333                  0.02                 0.01  198   
334                  0.02                 0.01  198   
336                  0.03                 0.02  198   
340                  0.07                 0.04  198   
341                  0.08                 0.05  198   

     Cost for Linux On Demand for 1 Year  Cost for Linux Reserved for 1 Year  
13                                 864.0                               518.4  
17                                 864.0                               604.8  
19                                1123.2                               864.0  
21                                 777.6                               518.4  
24                                 6

In [31]:
recommendations_1['Cost Saved on Linux Demand']=forecasted_grouped_data[forecasted_grouped_data['VM']==1].loc[:,'Forecasted Cost of OnDemand Linux Service for 1 Year'].values-recommendations_1.loc[:,'Cost for Linux On Demand for 1 Year'].values
recommendations_1['Cost Saved on Linux Reserved']=forecasted_grouped_data[forecasted_grouped_data['VM']==1].loc[:,'Forecasted Cost of Reserved Linux Service for 1 Year'].values-recommendations_1.loc[:,'Cost for Linux Reserved for 1 Year'].values

C:\Users\chat2\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\chat2\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [32]:
print(forecasted_grouped_data[forecasted_grouped_data['VM']==1].loc[:,'Forecasted Cost of OnDemand Linux Service for 1 Year'].values.shape)
print(recommendations_1.loc[:,'Cost for Linux On Demand for 1 Year'].values.shape)

(1,)
(3,)


In [33]:
recommendations_3['Cost Saved on Linux Demand']=forecasted_grouped_data[forecasted_grouped_data['VM']==3].loc[:,'Forecasted Cost of OnDemand Linux Service for 1 Year'].values-recommendations_3.loc[:,'Cost for Linux On Demand for 1 Year'].values
recommendations_3['Cost Saved on Linux Reserved']=forecasted_grouped_data[forecasted_grouped_data['VM']==3].loc[:,'Forecasted Cost of Reserved Linux Service for 1 Year'].values-recommendations_3.loc[:,'Cost for Linux Reserved for 1 Year'].values

C:\Users\chat2\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\chat2\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [34]:
recommendations_198['Cost Saved on Linux Demand']=forecasted_grouped_data[forecasted_grouped_data['VM']==198].loc[:,'Forecasted Cost of OnDemand Linux Service for 1 Year'].values[0]-recommendations_198.loc[:,'Cost for Linux On Demand for 1 Year'].values
recommendations_198['Cost Saved on Linux Reserved']=forecasted_grouped_data[forecasted_grouped_data['VM']==198].loc[:,'Forecasted Cost of Reserved Linux Service for 1 Year'].values[0]-recommendations_198.loc[:,'Cost for Linux Reserved for 1 Year'].values

C:\Users\chat2\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\chat2\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [35]:

frames_rec = [recommendations_1, recommendations_3, recommendations_198]

recommendations_final = pd.concat(frames_rec)

In [36]:
recommendations_final.to_csv('D:/Docs_copy/Capstone/Resource Data/recommendations.csv')

In [37]:
f=pd.read_csv('D:/Docs_copy/Capstone/Resource Data/recommendations.csv')

In [38]:
f.head()

Unnamed: 0                      Name   API Name  Memory  Cores  \
0           8                 A1 Medium  a1.medium     2.0      1   
1          13  M4 General Purpose Large   m4.large     8.0      2   
2          34                  A1 Large   a1.large     4.0      2   
3           8                 A1 Medium  a1.medium     2.0      1   
4          13  M4 General Purpose Large   m4.large     8.0      2   

                Physical Processor  Clock Speed(GHz)  Linux On Demand cost  \
0           AWS Graviton Processor               2.3                  0.03   
1  Intel Xeon E5-2676 v3 (Haswell)               2.4                  0.10   
2           AWS Graviton Processor               2.3                  0.05   
3           AWS Graviton Processor               2.3                  0.03   
4  Intel Xeon E5-2676 v3 (Haswell)               2.4                  0.10   

   Linux Reserved cost   VM  Cost for Linux On Demand for 1 Year  \
0                 0.02    1                                259.2   
1                 0.06    1                                864.0   
2                 0.03    1                                432.0   
3                 0.02    3                                259.2   
4                 0.06  198                                864.0   

   Cost for Linux Reserved for 1 Year  Cost Saved on Linux Demand  \
0                               172.8                      915.84   
1                               518.4                      311.04   
2                               259.2                      743.04   
3                               172.8                      630.72   
4                               518.4                      -86.40   

   Cost Saved on Linux Reserved  
0                        518.40  
1                        172.80  
2                        432.00  
3                        388.80  
4                        -25.92

In [39]:
forecasted_grouped_data.head()

Model   VM  CPU_MAE  CPU_RMSE  MPU_MAE  MPU_RMSE  CPU cores  \
0    AutoRegressive    3   0.0107   0.03212  5.56943   2.35996          2   
1  AutoRegressive_2  198   0.4340   0.59020  1.28500   1.64150          2   
2        FB Prophet    1   1.6741   2.40698  2.97718   3.98274          2   
3    Moving Average  198   0.4406   0.59730  1.28500   1.64150          2   

   CPU capacity provisioned [MHZ]  predicted_cpu_usage [%]  CPU usage [MHZ]  \
0                     2599.999304                 0.994962        25.869017   
1                     5199.999272                50.495910      2625.786931   
2                     5851.998900                 9.223119       539.736850   
3                     5199.999272                50.492804      2625.625428   

   Memory capacity provisioned [KB]  predicted_memory_usage [%]  \
0                            702948                   10.515294   
1                           8257536                   17.294269   
2                           8218624                   13.752128   
3                           8257536                   17.295391   

   Memory usage [KB]  CPU Provisoned  Memory Provisoned  \
0       7.391705e+04             2.5                  8   
1       1.428081e+06             2.5                  8   
2       1.130236e+06             3.0                  8   
3       1.428173e+06             2.5                  8   

   Cost of OnDemand Linux Service  Cost of Reserved Linux Service  \
0                          222.48                          140.40   
1                          194.40                          123.12   
2                          293.76                          172.80   
3                          194.40                          123.12   

   Forecasted Cost of OnDemand Linux Service for 1 Year  \
0                                             889.92      
1                                             777.60      
2                                            1175.04      
3                                             777.60      

   Forecasted Cost of Reserved Linux Service for 1 Year  
0                                             561.60     
1                                             492.48     
2                                             691.20     
3                                             492.48

In [40]:
recommendations_final

Name    API Name  Memory  Cores  \
8                   A1 Medium   a1.medium     2.0      1   
13   M4 General Purpose Large    m4.large     8.0      2   
34                   A1 Large    a1.large     4.0      2   
8                   A1 Medium   a1.medium     2.0      1   
13   M4 General Purpose Large    m4.large     8.0      2   
17                 M5AD Large  m5ad.large     8.0      2   
19   M3 General Purpose Large    m3.large     7.5      2   
21                 M6GD Large  m6gd.large     8.0      2   
24                 C6GD Large  c6gd.large     4.0      2   
32                  M5A Large   m5a.large     8.0      2   
34                   A1 Large    a1.large     4.0      2   
35                  C6G Large   c6g.large     4.0      2   
326                  T3A Nano    t3a.nano     0.5      2   
327                  T4G Nano    t4g.nano     0.5      2   
329                 T4G Micro   t4g.micro     1.0      2   
330                 T3A Micro   t3a.micro     1.0      2   
332                T3A Medium  t3a.medium     4.0      2   
333                 T4G Small   t4g.small     2.0      2   
334                 T3A Small   t3a.small     2.0      2   
336                T4G Medium  t4g.medium     4.0      2   
340                 T4G Large   t4g.large     8.0      2   
341                 T3A Large   t3a.large     8.0      2   

                                  Physical Processor  Clock Speed(GHz)  \
8                             AWS Graviton Processor               2.3   
13                   Intel Xeon E5-2676 v3 (Haswell)               2.4   
34                            AWS Graviton Processor               2.3   
8                             AWS Graviton Processor               2.3   
13                   Intel Xeon E5-2676 v3 (Haswell)               2.4   
17                                     AMD EPYC 7571               2.5   
19   Intel Xeon E5-2670 v2 (Ivy Bridge/Sandy Bridge)               2.5   
21                           AWS Graviton2 Processor               2.5   
24                           AWS Graviton2 Processor               2.5   
32                                     AMD EPYC 7571               2.5   
34                            AWS Graviton Processor               2.3   
35                           AWS Graviton2 Processor               2.5   
326                                    AMD EPYC 7571               2.5   
327                          AWS Graviton2 Processor               2.5   
329                          AWS Graviton2 Processor               2.5   
330                                    AMD EPYC 7571               2.5   
332                                    AMD EPYC 7571               2.5   
333                          AWS Graviton2 Processor               2.5   
334                                    AMD EPYC 7571               2.5   
336                          AWS Graviton2 Processor               2.5   
340                          AWS Graviton2 Processor               2.5   
341                                    AMD EPYC 7571               2.5   

     Linux On Demand cost  Linux Reserved cost   VM  \
8                    0.03                 0.02    1   
13                   0.10                 0.06    1   
34                   0.05                 0.03    1   
8                    0.03                 0.02    3   
13                   0.10                 0.06  198   
17                   0.10                 0.07  198   
19                   0.13                 0.10  198   
21                   0.09                 0.06  198   
24                   0.08                 0.05  198   
32                   0.09                 0.05  198   
34                   0.05                 0.03  198   
35                   0.07                 0.04  198   
326                  0.00                 0.00  198   
327                  0.00                 0.00  198   
329                  0.01                 0.01  198   
330                  0.01                 0.01  198   
332   

In [43]:
forecasted_grouped_data['PercentCount']=100

In [45]:
forecasted_grouped_data.to_csv('D:/Docs_copy/Capstone/Resource Data/ForecastedCostVM.csv')

In [44]:
forecasted_grouped_data.head()

Model   VM  CPU_MAE  CPU_RMSE  MPU_MAE  MPU_RMSE  CPU cores  \
0    AutoRegressive    3   0.0107   0.03212  5.56943   2.35996          2   
1  AutoRegressive_2  198   0.4340   0.59020  1.28500   1.64150          2   
2        FB Prophet    1   1.6741   2.40698  2.97718   3.98274          2   
3    Moving Average  198   0.4406   0.59730  1.28500   1.64150          2   

   CPU capacity provisioned [MHZ]  predicted_cpu_usage [%]  CPU usage [MHZ]  \
0                     2599.999304                 0.994962        25.869017   
1                     5199.999272                50.495910      2625.786931   
2                     5851.998900                 9.223119       539.736850   
3                     5199.999272                50.492804      2625.625428   

   Memory capacity provisioned [KB]  predicted_memory_usage [%]  \
0                            702948                   10.515294   
1                           8257536                   17.294269   
2                           8218624                   13.752128   
3                           8257536                   17.295391   

   Memory usage [KB]  CPU Provisoned  Memory Provisoned  \
0       7.391705e+04             2.5                  8   
1       1.428081e+06             2.5                  8   
2       1.130236e+06             3.0                  8   
3       1.428173e+06             2.5                  8   

   Cost of OnDemand Linux Service  Cost of Reserved Linux Service  \
0                          222.48                          140.40   
1                          194.40                          123.12   
2                          293.76                          172.80   
3                          194.40                          123.12   

   Forecasted Cost of OnDemand Linux Service for 1 Year  \
0                                             889.92      
1                                             777.60      
2                                            1175.04      
3                                             777.60      

   Forecasted Cost of Reserved Linux Service for 1 Year  PercentCount  
0                                             561.60              100  
1                                             492.48              100  
2                                             691.20              100  
3                                             492.48              100